openmls/src/treesync/node/encryption_keys.rs

```diff
@@ -0,0 +1,189 @@
+        backend.key_store().store(&key_store_index, self)
+    }
+
+    /// Read the [`EncryptionKeyPair`] from the key store of the `backend`. This
+    /// function is meant to read standalone keypairs, not ones that are
+    /// already in use with an MLS group.
+    ///
+    /// Returns `None` if the keypair cannot be read from the store.
+    pub fn read_from_key_store(
+        backend: &impl OpenMlsCryptoProvider,
+        encryption_key: &EncryptionKey,
+    ) -> Option<EncryptionKeyPair> {
+        let mut key_store_index = ENCRYPTION_KEY_LABEL.to_vec();
+        key_store_index.extend_from_slice(encryption_key.as_slice());
+        backend.key_store().read(&key_store_index)
+    }
+
+    /// Delete the [`EncryptionKeyPair`] from the key store of the `backend`.
+    /// This function is meant to delete standalone keypairs, not ones that are
+    /// already in use with an MLS group.
+    ///
+    /// Returns a key store error if access to the key store fails.
+    pub fn delete_from_key_store<KeyStore: OpenMlsKeyStore>(
+        &self,
+        backend: &impl OpenMlsCryptoProvider<KeyStoreProvider = KeyStore>,
+    ) -> Result<(), KeyStore::Error> {
+        let mut key_store_index = ENCRYPTION_KEY_LABEL.to_vec();
+        key_store_index.extend_from_slice(self.public_key().as_slice());
+        backend.key_store().delete(&key_store_index)
+    }
+
+    pub fn public_key(&self) -> &EncryptionKey {
+        &self.public_key
+    }
+
+    pub fn private_key(&self) -> &EncryptionPrivateKey {
+        &self.private_key
+    }
+
+    pub(crate) fn derive(
+        backend: &impl OpenMlsCryptoProvider,
+        ciphersuite: Ciphersuite,
+        ikm: Secret,
+    ) -> Self {
+        backend
+            .crypto()
+            .derive_hpke_keypair(ciphersuite.hpke_config(), ikm.as_slice())
+            .into()
+    }
+}
+
+impl From<(HpkePublicKey, HpkePrivateKey)> for EncryptionKeyPair {
+    fn from((public_key, private_key): (HpkePublicKey, HpkePrivateKey)) -> Self {
+        Self {
+            public_key: public_key.into(),
+            private_key: private_key.into(),
+        }
+    }
+}
+
+impl From<HpkeKeyPair> for EncryptionKeyPair {
+    fn from(hpke_keypair: HpkeKeyPair) -> Self {
+        let public_bytes: VLBytes = hpke_keypair.public.into();
+        let private_bytes: VLBytes = hpke_keypair.private.into();
+        Self {
+            public_key: public_bytes.into(),
+            private_key: private_bytes.into(),
+        }
+    }
+}
+
+impl From<(EncryptionKey, EncryptionPrivateKey)> for EncryptionKeyPair {
+    fn from((public_key, private_key): (EncryptionKey, EncryptionPrivateKey)) -> Self {
+        Self {
+            public_key,
+            private_key,
+        }
+    }
+}
+
+impl ToKeyStoreValue for EncryptionKeyPair {
+    type Error = LibraryError;
+
+    fn to_key_store_value(&self) -> Result<Vec<u8>, Self::Error> {
+        serde_json::to_vec(self)
```

Comment:
I tried refactoring to avoid deriving the serde traits, but we have `EncryptionKeyPair` in the `MemberStagedCommitState`, which we need to serialize :-(

Reply:
Yes, I filed #1194 to make track how persistence is coordinated.